cub/cub/device/dispatch/dispatch_merge.cuh

```diff
@@ -0,0 +1,351 @@
+  MergeAgent{
+    temp_storage.Alias(),
+    make_load_iterator(MergePolicy{}, keys1),
+    make_load_iterator(MergePolicy{}, items1),
+    num_keys1,
+    make_load_iterator(MergePolicy{}, keys2),
+    make_load_iterator(MergePolicy{}, items2),
+    num_keys2,
+    keys_result,
+    items_result,
+    compare_op,
+    merge_partitions}();
+  vsmem_helper_t::discard_temp_storage(temp_storage);
+}
+
+template <typename KeyT, typename ValueT>
+struct device_merge_policy_hub
+{
+  static constexpr bool has_values = !::cuda::std::is_same<ValueT, NullType>::value;
+
+  // TODO(bgruber): is this correct?
+  using tune_type = char[has_values ? sizeof(KeyT) + sizeof(ValueT) : sizeof(KeyT)];
+
+  struct policy300 : ChainedPolicy<300, policy300, policy300>
+  {
+    using merge_policy =
+      agent_policy_t<128,
+                     Nominal4BItemsToItems<tune_type>(7),
+                     BLOCK_LOAD_WARP_TRANSPOSE,
+                     LOAD_DEFAULT,
+                     BLOCK_STORE_WARP_TRANSPOSE>;
+  };
+
+  struct policy350 : ChainedPolicy<350, policy350, policy300>
+  {
+    using merge_policy =
+      agent_policy_t<256,
+                     Nominal4BItemsToItems<tune_type>(11),
+                     BLOCK_LOAD_WARP_TRANSPOSE,
+                     LOAD_LDG,
+                     BLOCK_STORE_WARP_TRANSPOSE>;
+  };
+
+  struct policy520 : ChainedPolicy<520, policy520, policy350>
+  {
+    using merge_policy =
+      agent_policy_t<512,
+                     Nominal4BItemsToItems<tune_type>(13),
+                     BLOCK_LOAD_WARP_TRANSPOSE,
+                     LOAD_LDG,
+                     BLOCK_STORE_WARP_TRANSPOSE>;
+  };
+
+  struct policy600 : ChainedPolicy<600, policy600, policy520>
+  {
+    using merge_policy =
+      agent_policy_t<512,
+                     Nominal4BItemsToItems<tune_type>(15),
+                     BLOCK_LOAD_WARP_TRANSPOSE,
+                     LOAD_DEFAULT,
+                     BLOCK_STORE_WARP_TRANSPOSE>;
+  };
+
+  using max_policy = policy600;
+};
+
+template <typename KeyIt1,
+          typename ValueIt1,
+          typename KeyIt2,
+          typename ValueIt2,
+          typename KeyIt3,
+          typename ValueIt3,
+          typename Offset,
+          typename CompareOp,
+          typename PolicyHub = device_merge_policy_hub<value_t<KeyIt1>, value_t<ValueIt1>>>
+struct dispatch_t
+{
+  using key_t   = cub::detail::value_t<KeyIt1>;
+  using value_t = cub::detail::value_t<ValueIt1>;
+
+  // Cannot check output iterators, since they could be discard iterators, which do not have the right value_type
+  static_assert(::cuda::std::is_same<cub::detail::value_t<KeyIt2>, key_t>::value, "");
+  static_assert(::cuda::std::is_same<cub::detail::value_t<ValueIt2>, value_t>::value, "");
+  static_assert(::cuda::std::__invokable<CompareOp, key_t, key_t>::value,
+                "Comparison operator cannot compare two keys");
```

Comment:
suggestion: it's not required, but we try to make errors more user-friendly with tag dispatch in the API entry point: https://github.com/NVIDIA/cccl/blob/20cd6ce34250acc7e33df584dbb196ba0526d86a/cub/cub/device/device_radix_sort.cuh#L522-L536

Reply:
I don't understand what you mean by tag dispatch here. And you are suggesting duplicating the above checks to all API entry points? And how is user friendliness increased? Is the goal to give them a smaller template instantiation stack when the error occurs?

Reply:
> I don't understand what you mean by tag dispatch here.

Here's an introductory [material](https://iamsorush.com/posts/cpp-tag-dispatch/) into tag dispatch. If your question is on how to apply tag dispatch in regards to static asserts, you can take a look at the reference above. The idea is to wrap condition into integral constant and have two overloads accepting this constant as a tag.

> And you are suggesting duplicating the above checks to all API entry points? 

It's optional, but yes, you can extract the check into a type alias and reuse it across two new API entry points.

> And how is user friendliness increased? Is the goal to give them a smaller template instantiation stack when the error occurs?

Yes, the goal is to avoid pages of compiler errors and print error text with no extra compiler output.

Reply:
> Here's an introductory [material](https://iamsorush.com/posts/cpp-tag-dispatch/) into tag dispatch.

I understand what tag dispatch is.

> If your question is on how to apply tag dispatch in regards to static asserts, you can take a look at the reference above. The idea is to wrap condition into integral constant and have two overloads accepting this constant as a tag.

So what you want seems to be the following:

* move the `static_assert` from the dispatch layer to the API entry points, so the template instantiation stack is one level shallower if the assert fires. But IIUC, this has nothing to do with tag dispatch.
* avoid the call to the dispatch layer if the `static_assert` fires, so the user does not see any follow-up errors? which is implemented using tag dispatch in your example, but could just be an `if constexpr` as well.

> It's optional, but yes, you can extract the check into a type alias and reuse it across two new API entry points.

I would need to use it in three entry points, because I want the asserts to fire as well when somebody calls directly into the dispatch layer, as the unit tests and Thrust do (the latter because of the offset type workaround).

> Yes, the goal is to avoid pages of compiler errors and print error text with no extra compiler output.

The current solution causes one extra level of template instantiation shown by the compiler because of diagnosing in the dispatch layer rather than in the API entrypoint. Showing follow-up errors after a `static_assert` fires is a choice the compiler makes and can also be changed by passing different compilation flags (e.g. `-Wfatal-errors`). Duplicating the check adds more redundancy to the code base as well.

I would like to politely refuse to apply this suggestion.

Reply:
I just had another go at it and would like to stay with my position :)